crates/ruff_linter/src/rules/flake8_type_checking/helpers.rs

```diff
@@ -14,35 +14,110 @@ use ruff_text_size::{Ranged, TextRange};
 
 use crate::Edit;
 use crate::Locator;
-use crate::rules::flake8_type_checking::settings::Settings;
-
-/// Returns `true` if the [`ResolvedReference`] is in a typing-only context _or_ a runtime-evaluated
-/// context (with quoting enabled).
-pub(crate) fn is_typing_reference(reference: &ResolvedReference, settings: &Settings) -> bool {
-    reference.in_type_checking_block()
-        // if we're not in a type checking block, we necessarily need to be within a
-        // type definition to be considered a typing reference
-        || (reference.in_type_definition()
-            && (reference.in_typing_only_annotation()
-                || reference.in_string_type_definition()
-                || (settings.quote_annotations && reference.in_runtime_evaluated_annotation())))
+use crate::settings::LinterSettings;
+
+/// Represents the kind of an existing or potential typing-only annotation.
+///
+/// See the `from_references` constructor for more details.
+#[derive(Clone, Copy)]
+pub(crate) enum TypingReference {
+    Yes,
+    No,
+    Quote,
+    Future,
+}
+
+impl TypingReference {
+    /// Determine the kind of [`TypingReference`] for a sequence of [`ResolvedReference`]s.
+    ///
+    /// This can be used to determine if all of the references to a binding are in a typing-only
+    /// context (`TypingReference::Yes`), in a runtime-evaluated context with
+    /// `lint.allow_importing_future_annotations` enabled (`TypingReference::Future`), or in a
+    /// runtime-evaluated context with `lint.flake8_type_checking.quote_annotations` enabled
+    /// (`TypingReference::Quote`).
```

Comment:
nit: I'd find this documentation easier to read if you had a separate doc-comment for each variant above, I think